README.md
[![GoDoc Reference](http://img.shields.io/badge/godoc-reference-5272B4.svg?style=flat-square)](http://godoc.org/github.com/FATHOM5/haversine)
[![Build Status](https://travis-ci.org/FATHOM5/haversine.svg?branch=master)](https://travis-ci.org/FATHOM5/haversine)
[![Go Report Card](https://goreportcard.com/badge/github.com/FATHOM5/haversine?style=flat-square)](https://goreportcard.com/report/github.com/FATHOM5/haversine)
[![Coverage Status](https://coveralls.io/repos/github/FATHOM5/haversine/badge.svg?branch=master)](https://coveralls.io/github/FATHOM5/haversine?branch=master)

## Haversine

Package haversine provides the great circle distance between two points on the surface of a sphere, with specific functions for Earth.  Points are identified by latitude and longitude, and distance results can be returned in nautical miles, statute miles, or kilometers. An internal angle function returns the radians between two points.
![Earth great circle](https://i.imgur.com/iD3X3Ax.png)

The curvature of the Earth dictates that the shortest distance between two points cannot be a straight line between the points.  The Haversine formula provides an approximation for the shortest great circle route over the surface of Earth that connects the two points.  In this figure the dotted yellow line is the arc of a great circle. *Image courtesy USGS.*

## Installation

`go get github.com/FATHOM5/haversine`

## Usage

The example below shows how far I'd have to walk in my boots to go visit my
co-founder.

    package main

    import (
        "fmt"
        "math"

        "github.com/FATHOM5/haversine"
    )

    func main() {
        austin := haversine.Coord{ // Austin, Texas
            Lat: 30.2672,
            Lon: -97.7431,
        }
        paloAlto := haversine.Coord{ // Palo Alto, California
            Lat: 37.4419,
            Lon: -122.1430,
        }

        nm := haversine.DistanceNM(austin, paloAlto)

        fmt.Printf("%.1f miles is a long walk to Silicon Valley.\n", nm)
        fmt.Printf("it's only %.1f miles on the moon, though.\n", haversine.Distance(austin, paloAlto, 937.9))
        fmt.Printf("That's an angle of %.1f degrees\n", 180/math.Pi*haversine.IntAngle(austin, paloAlto))

        // Output: 1286.1 miles is a long walk to Silicon Valley.
        // It's only 350.6 miles on the moon, though.
        // That's an angle of 21.4 degrees

    }

## Documentation

* http://godoc.org/github.com/FATHOM5/haversine

## References

* https://plus.maths.org/content/lost-lovely-haversine
* https://en.wikipedia.org/wiki/Haversine_formula
* Many thanks to the original author, [umahmood](https://github.com/umahmood/haversine), for a great package to begin this fork.

## License

See [LICENSE](LICENSE.md) for rights and limitations (MIT).


CHANGELOG.md
# Change Log
All notable changes documented in adherance to [Semantic Versioning](http://semver.org/).

Prior to June 2020 the API for this package and it's parent at umahmood/haversine had major version number changes for each  adaptation of the API.  In June 2020, the original 2015 API was renumbered to v0.1.0 to reflect that it was premature to declare a v1.0.0 release without a larger user community for this small package. Edits to the package in 2018 by Fathom5 and open source contributions in 2020 were considered minor API version changes prior to an official 1.0.0 release that will occur when the package is used by at least ten other projects or contributors.

This change will better reflect the package's maturity in the [golang package ecosystem](https://pkg.go.dev/)

v0.1.0 - Original API last committed in Nov 2015

v0.2.0 - API changes committed in Nov 2018.  See commit [092bd4b](https://github.com/FATHOM5/haversine/commit/092bd4b5508f4355bafeb2d632eded2cf958a251)

v0.3.0 - API changes committed in May 2020. See commit [](https://github.com/FATHOM5/haversine/commit/.......)
  Refactored Distance to use a third input param, the radius of the sphere, to create a more generalized function.
  Created DistanceNM(p, q Coord) function to return nautical miles on Earth.
  Created IntAngle(p, q Coord) function to return radians between two points. Above Distance function calls this.
  Exported miles, kilometer and nautical mile Earth radius values
  Exported distance conversion functions
  Fixed errors in haversine_test.go


planet_test.go
package haversine_test

import (
        "fmt"

        "github.com/FATHOM5/haversine"
)

func main(){
        // a planety thing
        type planet struct {
                name string
                r float64
        }

        // km radius
        var solarSystem = []planet {
                {"the Sun",695700},
                {"the Moon",1737},
                {"Mercury",2440},
                {"Venus",6051},
                {"Earth",6371},
                {"Mars",3390},
                {"Jupiter",69910},
                {"Saturn",58230},
                {"Uranus",25360},
                {"Neptune",24620},
                {"the unit sphere",1},
                // Sorry, Pluto, you got kicked to the curb, dude.
                // Your sorry ass is even smaller than the moon, anyway, so,
                // "Meh" is what I say to you. Sorry, Clyde. I love you, man.
        }

        // a winter seaside resort town
        wilbur := haversine.Coord{29.13, -80.97}
        fmt.Println("Wilbur-by-the-Sea is here:", wilbur)

        b := wilbur.Radians()
        fmt.Println(" which, in radians, is", b)
        fmt.Println(" and inverted back to degrees yields", b.Degrees())

        // it's antipode, a not winter not seaside not resort town
        summerhills := haversine.Coord{45.71, -122.43}
        fmt.Println("Summer Hills, however, is here:", summerhills)

        // regardless of sphere, what's the angle?
        fmt.Printf("That's an angular distance of %.4f radians\n", haversine.IntAngle(wilbur, summerhills))
        fmt.Printf(" or %.2f degrees\n", 180 / 3.14159 * haversine.IntAngle(wilbur, summerhills))
        fmt.Printf(" so from Wilbur to Summer Hills is %.0f miles.\n", haversine.DistanceMi(wilbur, summerhills))

        // if this were a solar system object
        for _, v := range solarSystem {
                fmt.Printf("On %s, that would be %.1f miles or %.1f days of driving\n", v.name, haversine.Distance(wilbur, summerhills, v.r / 1.6093), 6*v.r/6371)
        }

}


example_test.go
// Example demonstrates use of the haversine Distance function.
package haversine_test

import (
	"fmt"

	"github.com/FATHOM5/haversine"
)

func ExampleDistanceNM() {
	austin := haversine.Coord{ // Austin, Texas
		Lat: 30.2672,
		Lon: -97.7431,
	}
	paloAlto := haversine.Coord{ // Palo Alto, California
		Lat: 37.4419,
		Lon: -122.1430,
	}

	nm := haversine.DistanceNM(austin, paloAlto)

	fmt.Printf("%.1f miles is a long walk to Silicon Valley.\n", nm)

	// Output: 1286.1 miles is a long walk to Silicon Valley.
}

func ExampleDistance() {
	austin := haversine.Coord{ // Austin, Texas
		Lat: 30.2672,
		Lon: -97.7431,
	}
	paloAlto := haversine.Coord{ // Palo Alto, California
		Lat: 37.4419,
		Lon: -122.1430,
	}

	km := haversine.Distance(austin, paloAlto, 58230)

	fmt.Printf("%.1f km is a long walk to Silicon Valley on Saturn.\n", km)

	// Output: 21770.2 km is a long walk to Silicon Valley on Saturn.
}

func ExampleIntAngle() {
	austin := haversine.Coord{ // Austin, Texas
		Lat: 30.2672,
		Lon: -97.7431,
	}
	paloAlto := haversine.Coord{ // Palo Alto, California
		Lat: 37.4419,
		Lon: -122.1430,
	}

	radians := haversine.IntAngle(austin, paloAlto)

	fmt.Printf("%.4f radians is a long walk to Silicon Valley, perhaps.\n", radians)

	// Output: 0.3739 radians is a long walk to Silicon Valley, perhaps.
}

func ExampleNmToKm() {
	fmt.Printf("One degree is 60 NM which is %.1f kilometers\n", haversine.NmToKm(60))

	// Output: One degree is 60 NM which is 111.1 kilometers
}

func ExampleDegrees() {
	// a planety thing
	type planet struct {
		name string
		r    float64
	}

	// km radius
	var solarSystem = []planet{
		{"the Sun", 695700},
		{"the Moon", 1737},
		{"Mercury", 2440},
		{"Venus", 6051},
		{"Earth", 6371},
		{"Mars", 3390},
		{"Jupiter", 69910},
		{"Saturn", 58230},
		{"Uranus", 25360},
		{"Neptune", 24620},
		{"the unit sphere", 1},
		// Sorry, Pluto, you got kicked to the curb, dude.
		// Your sorry ass is even smaller than the moon, anyway, so,
		// "Meh" is what I say to you. Sorry, Clyde. I love you, man.
	}

	// a winter seaside resort town
	wilbur := haversine.Coord{29.13, -80.97}
	fmt.Println("Wilbur-by-the-Sea is here:", wilbur)

	b := wilbur.Radians()
	fmt.Println(" which, in radians, is", b)
	fmt.Println(" and inverted back to degrees yields", b.Degrees())

	// it's antipode, a not winter not seaside not resort town
	summerhills := haversine.Coord{45.71, -122.43}
	fmt.Println("Summer Hills, however, is here:", summerhills)

	// regardless of sphere, what's the angle?
	fmt.Printf(" That's an angular distance of %.4f radians\n", haversine.IntAngle(wilbur, summerhills))
	fmt.Printf(" or %.2f degrees\n", 180/3.14159*haversine.IntAngle(wilbur, summerhills))
	fmt.Printf(" so from Wilbur to Summer Hills is %.0f miles.\n", haversine.DistanceMi(wilbur, summerhills))

	// if this were a solar system object
	for _, v := range solarSystem {
		fmt.Printf("On %s, that would be %.1f miles or %.1f days of driving\n", v.name, haversine.Distance(wilbur, summerhills, haversine.NmToMi(haversine.KmToNm(v.r))), 6*v.r/6371)
	}

	// Output:
	// Wilbur-by-the-Sea is here: {29.13 -80.97}
	//  which, in radians, is {0.5084144111059482 -1.4131930953398086}
	//  and inverted back to degrees yields {29.130000000000003 -80.97}
	// Summer Hills, however, is here: {45.71 -122.43}
	//  That's an angular distance of 0.6342 radians
	//  or 36.33 degrees
	//  so from Wilbur to Summer Hills is 2510 miles.
	// On the Sun, that would be 274083.4 miles or 655.2 days of driving
	// On the Moon, that would be 684.3 miles or 1.6 days of driving
	// On Mercury, that would be 961.3 miles or 2.3 days of driving
	// On Venus, that would be 2383.9 miles or 5.7 days of driving
	// On Earth, that would be 2510.0 miles or 6.0 days of driving
	// On Mars, that would be 1335.6 miles or 3.2 days of driving
	// On Jupiter, that would be 27542.3 miles or 65.8 days of driving
	// On Saturn, that would be 22940.7 miles or 54.8 days of driving
	// On Uranus, that would be 9991.0 miles or 23.9 days of driving
	// On Neptune, that would be 9699.5 miles or 23.2 days of driving
	// On the unit sphere, that would be 0.4 miles or 0.0 days of driving

}


haversine_test.go
package haversine_test

import (
        "math"
        "testing"

        "github.com/FATHOM5/haversine"
)

// Great circle calculations are really only valid for about four sig figs
// for Earth, anyway. Up this, (or should that be, "decrease this"?)
// for real spheres.
const ppm = 0.0001

var tests = []struct {
        p     haversine.Coord
        q     haversine.Coord
        outMi float64
        outKm float64
        outRa float64
        town  string
}{
        {
                haversine.Coord{Lat: -22.55, Lon: -43.12},  // Rio de Janeiro, Brazil
                haversine.Coord{Lat: 13.45, Lon: 100.28}, // Bangkok, Thailand
                9956,
                16026,
                2.515463,
                "Rio to Bangkok",
        },
        {
                haversine.Coord{Lat: -20.10, Lon: 57.30}, // Port Louis, Mauritius
                haversine.Coord{Lat: 0.57, Lon: 100.21}, // Padang, Indonesia
                3233,
                5205,
                0.817067,
                "Mauritius to Indonesia",
        },
        {
                haversine.Coord{Lat: 51.45, Lon: -1.15},  // Oxford, United Kingdom
                haversine.Coord{Lat: 41.54, Lon: 12.27}, // Vatican, City Vatican City
                933,
                1501,
                0.235736,
                "Oxford to the Vatican",
        },
        {
                // These are antipodes
                haversine.Coord{Lat: 32.30, Lon: -64.77}, // Bermuda
                haversine.Coord{Lat: -32.30, Lon: 115.23},  // Perth, Australia
                12434,                // earthRadiusMi * pi
                20015,                // earthRadiusKm * pi
                3.14159,        // pi
                "Bermuda to Perth",
        },
        {
                haversine.Coord{Lat: -22.34, Lon: 17.05}, // Windhoek, Namibia
                haversine.Coord{Lat: 51.56, Lon: 4.29},  // Rotterdam, Netherlands
                5163,
                8311,
                1.304548,
                "Namibia to the Netherlands",
        },
        {
                haversine.Coord{Lat: -63.24, Lon: -56.59}, // Esperanza, Argentina
                haversine.Coord{Lat: -8.50, Lon: 13.14},  // Luanda, Angola
                5068,
                8157,
                1.280482,
                "Argentina to Angola",
        },
        {
                haversine.Coord{Lat: 90.00, Lon: 0.00}, // North Pole
                haversine.Coord{Lat: 48.51, Lon: 2.21}, // Paris,  France
                2866,
                4613,
                0.724137,
                "Santa to Paris",
        },
        {
                haversine.Coord{Lat: -90.00, Lon: 0.00}, // South Pole
                haversine.Coord{Lat: 48.51, Lon: 2.21}, // Paris,  France
                9568,
                15401,
                2.417456,
                "Penguins to Paris",
        },
        {
                haversine.Coord{Lat: 45.04, Lon: 7.42},  // Turin, Italy
                haversine.Coord{Lat: 3.09, Lon: 101.42}, // Kuala Lumpur, Malaysia
                6261,
                10078,
                1.581873,
                "Turin to Malaysia",
        },
        {
                haversine.Coord{Lat: 45.71, Lon: -122.43},  // Hockinson, Washington, USA
                haversine.Coord{Lat: 29.13, Lon: -80.96},   // Wilbur-by-the-Sea, Florida, USA
                2510,
                4040,
                0.634270,
                "Washington to Florida",
        },
}

// TestHaversineDistance makes sure that the several great circle distance
// functions work properly
func TestHaversineDistance(t *testing.T) {
        for _, input := range tests {
                mi, _ := math.Modf(haversine.DistanceMi(input.p, input.q))
                km, _ := math.Modf(haversine.DistanceKm(input.p, input.q))

                if input.outMi != mi || input.outKm != km {
                        t.Errorf("fail: want %v %v -> %v %v got %g %g %s",
                                input.p,
                                input.q,
                                input.outMi,
                                input.outKm,
                                mi,
                                km,
                                input.town,
                        )
                }
        }
}

// TestIntAngle makes sure the internal angle calculation result is within tolerance
func TestIntAngle(t *testing.T) {
        for _, input := range tests {
                iAngle := haversine.IntAngle(input.p, input.q)
                delta  := math.Abs(input.outRa - iAngle) / input.outRa

                if delta > ppm {
                        t.Errorf("fail: want %v %v -> %v got %.6f, %.1f ppm %s",
                                input.p,
                                input.q,
                                input.outRa,
                                iAngle,
                                delta*1000000,
                                input.town,
                        )
                }
        }
}

// TestPoleToPole makes sure the sum of internal angles of the north pole, a
// point and the south pole adds up to 180 degrees (pi radians)
func TestPoleToPole(t *testing.T) {
        reindeer := haversine.Coord{90,0}
        penguins := haversine.Coord{-90,0}
        for _, input := range tests {
                nAngle := haversine.IntAngle(input.q, reindeer)
                sAngle := haversine.IntAngle(input.q, penguins)
                tAngle := nAngle + sAngle
                delta  := math.Abs(tAngle - math.Pi) / math.Pi

                if delta > ppm {
                        t.Errorf("fail: want %v -> %-.5f + %-.5f = Pi got %-.5f, %.1f ppm %s",
                                input.q,
                                nAngle,
                                sAngle,
                                tAngle,
                                delta*1000000,
                                input.town,
                        )
                }
        }
}


haversine.go
package haversine

import (
	"math"
)

const (
	EarthRadiusMi = 3958 // radius of the earth in miles.
	EarthRadiusKm = 6371 // radius of the earth in kilometers.
	EarthRadiusNM = 3440 // radius of the earth in nautical miles
)

// Coord represents a lat/long geographic coordinate, usually in degrees +EN/-WS.
type Coord struct {
	Lat float64
	Lon float64
}

// Degrees converts a radians-based Coord to a degrees-based Coord
func (c *Coord) Degrees() Coord {
	return Coord{c.Lat * 180 / math.Pi, c.Lon * 180 / math.Pi}
}

// Radians converts a degrees-based Coord to a radians-based Coord
func (c *Coord) Radians() Coord {
	return Coord{c.Lat * math.Pi / 180, c.Lon * math.Pi / 180}
}

// NmToMi converts from nautical miles to statute miles.
func NmToMi(d float64) float64 {
	return d * EarthRadiusMi / EarthRadiusNM
}

// NmToKm converts from nautical miles to kilometers.
func NmToKm(d float64) float64 {
	return d * EarthRadiusKm / EarthRadiusNM
}

// KmToNm converts from kilometers to nautical miles.
func KmToNm(d float64) float64 {
	return d * EarthRadiusNM / EarthRadiusKm
}

// IntAgle calculates the internal angle between two coordinates on a surface
// and returns the result in radians
func IntAngle(p, q Coord) float64 {
	point1 := p.Radians()
	point2 := q.Radians()

	diffLat := point2.Lat - point1.Lat
	diffLon := point2.Lon - point1.Lon

	a := math.Pow(math.Sin(diffLat/2), 2) + math.Cos(point1.Lat)*math.Cos(point2.Lat)*
		math.Pow(math.Sin(diffLon/2), 2)

	return 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
}

// Distance calculates the shortest (aka great circle) arc between two coordinates on a sphere
// of a given radius and returns the resulting great circle arc length
func Distance(p, q Coord, radius float64) (gc float64) {

	c := IntAngle(p, q)

	return c * radius
}

// DistanceMi calculates the shortest path between two coordinates on the surface
// of the Earth and returns the result in statute miles.
func DistanceMi(p, q Coord) (mi float64) {
	return Distance(p, q, EarthRadiusMi)
}

// DistanceKm calculates the shortest path between two coordinates on the surface
// of the Earth and returns the result in kilometers.
func DistanceKm(p, q Coord) (km float64) {
	return Distance(p, q, EarthRadiusKm)
}

// DistanceNM calculates the shortest path between two coordinates on the surface
// of the Earth and returns the result in nautical miles.
func DistanceNM(p, q Coord) (nm float64) {
	return Distance(p, q, EarthRadiusNM)
}


doc.go
/*
Package haversine provides the great circle distance between two points
on a sphere. Several functions return great circle distances of the Earth.
Points are identified by latitude and longitude in degrees, and distance
results are returned in nautical miles.  Functions are also provided to
return the distance in kilometers or statute miles, or the geocentric
angular separation between those points.

The below example shows how to calculate the shortest path between two
coordinates on the surface of the Earth.

    package main

    import (
        "fmt"
        "math"

        "github.com/FATHOM5/haversine"
    )

    func main() {
        austin := haversine.Coord{ // Austin, Texas
            Lat: 30.2672,
            Lon: -97.7431,
        }
        paloAlto := haversine.Coord{ // Palo Alto, California
            Lat: 37.4419,
            Lon: -122.1430,
        }

        nm := haversine.DistanceNM(austin, paloAlto)

        fmt.Printf("%.1f miles is a long walk to Silicon Valley.\n", nm)
        fmt.Printf("it's only %.1f miles on the moon, though.\n", haversine.Distance(austin, paloAlto, 937.9))
        fmt.Printf("That's an angle of %.1f degrees\n", 180/math.Pi*haversine.IntAngle(austin, paloAlto))
    }
*/
package haversine
